_posts/2023-08-03-turbocharging-rust-code-verification.md

```diff
@@ -0,0 +1,445 @@
+Initially the *field sensitivity* transform in CBMC enabled constant propagation for individual array cells and individual struct fields, but not for union fields.
+Since constant propagation helps pruning control flow branches during symbolic execution and can greatly reduce the runtime of an analysis, ensuring that constant propagation also works for union fields is important for Rust programs.
+Field-sensitivity was first extended to unions in `cbmc-5.71.0`, but did not make it to Kani until `kani v0.17.0` built on top of `cbmc v5.72.0`.
+The feature was then refined and stabilized in several iterations and became stable with `cbmc v5.85.0` in early June 2023, and released through `kani v0.31.0` built on top of `cbmc v5.86.0`.
+This new CBMC feature vastly improved performance for Rust programs manipulating `Vec<T>` and `BTreeSet<T>` data types, and allowed us to solve a number performance issues reported by our users: [#705](https://github.com/model-checking/kani/issues/705), [#1226](https://github.com/model-checking/kani/issues/1226), [#1657](https://github.com/model-checking/kani/issues/1657), [#1673](https://github.com/model-checking/kani/issues/1673), [#1676](https://github.com/model-checking/kani/issues/1676).
+
+The following tables and plots were obtained by running the kani `perf` test suite with `kani 0.33.0`, `cbmc 5.88.1` with `cadical` as default SAT solver for all tests, a timeout of 900s, with and without applying the union-field sensitivity transform.
+
+
+| verification harness               | no-sens | sens | change          |
+| ---------------------------------- | ------- | ---- | --------------- |
+| btreeset/insert_any/main           | False   | True | ✅ newly passing |
+| btreeset/insert_multi/insert_multi | False   | True | ✅ newly passing |
+| btreeset/insert_same/main          | False   | True | ✅ newly passing |
+| misc/display_trait/slow            | False   | True | ✅ newly passing |
+| misc/struct_defs/fast_harness      | False   | True | ✅ newly passing |
+
+Field sensitivity allows 5 new verification harnesses to be solved under the 900s limit.
+
+![Total time](/kani-verifier-blog/assets/images/field-sens-plots/total-time.png)
+
+We observe significant 10% to 99% reduction in rutime for roughly a third of verification harnesses (and 5 newly solved harnesses within the 900s time limit), but we also observe a 10% to 55% runtime degradation or 25% of the verification harnesses, so this feature does not bring a consistent benefit. However, the cumulative total time to run the `perf` suite without union field sensitivity is roughly 6500s, and it drops to 1784s with union field sensitivity activated. Disabling union field sensitivity for harnesses where it degrades performance would only bring the cumulative total time down to 1676s. So we can say getting rid of timeouts and gains on a subset of harnesses offset the losses on the rest of harnesses.
```

Comment:
```suggestion
We observe a 10% to 99% reduction in runtime for roughly a third of verification harnesses (and 5 newly solved harnesses within the 900s time limit), but we also observe a 10% to 55% runtime degradation for 25% of the verification harnesses, so this feature does not bring a consistent benefit. However, the cumulative total time to run the `perf` suite without union field sensitivity is roughly 6500s, and it drops to 1784s with union field sensitivity activated. Disabling union field sensitivity for harnesses where it degrades performance would only bring the cumulative total time down to 1676s. So we can say getting rid of timeouts and gains on a subset of harnesses offset the losses on the rest of harnesses.
```

Reply:
Readers are likely not familiar with the "perf suite". Perhaps change to "on Kani's performance benchmarking suite"?

Reply:
The last two sentences could be confusing to the reader, and I'm not clear on the purpose they're serving. I suggest removing them.

Reply:
link added

Reply:
sentences removed